runtime/src/accounts_db.rs

```diff
@@ -3660,7 +3929,11 @@ impl AccountsDb {
                                 {
                                     let (slot, account_info) = &lock.slot_list()[index];
                                     if account_info.lamports != 0 {
-                                        self.get_account_accessor_from_cache_or_storage(
+                                        // Because we're keeing the `lock' here, there is no need
+                                        // to use retry_to_get_account_accessor()
+                                        // In other words, flusher/shrinker/cleaner is blocked to
+                                        // cause any Accessor(None) situtation.
```

Comment:
I hope my reasoning is right here?

Reply:
1) It should be safe with cache flush because cache flush `AccountsDb::flush_slot()`doesn't remove the slot entry from the cache until the accounts index gets updated first, which can't happen while we hold the lock here.

2) Any reclaims in `clean_accounts()` needs to go though the accounts index lock: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L1493-L1494 first, so the slot can't be dead yet, and the slot storages should not be cleaned up by `process_dead_slots()` yet

3) In shrink, this logic which updates the accounts index: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L1908-L1914 needs to run first before storage entries are removed: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L1926-L1930, so this should also be ok

Moreover for 2) and 3) loads in `calculate_accounts_hash()` should not  *currently* race with clean/shrink because the full hash is synchronous with clean/shrink in AccountsBackgroundService, but this could change in the future, so it's good you called it out here 😄

Reply:
cool! thanks for confirming. :)